File: Robot/Java/Helper/Command.java
package Robot.Java.Helper;

import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * A class which can send and read Commands over a OutputStream/InputStream
 * 
 * It contains a clientID, commandString and a HashMap of parameters
 */
public class Command {
	
	public static final String endDelimiter = "%%%%";
	
	/**
	 * @param string which gets checked
	 * @return if string contains one of the delimiters it returns true otherwise false
	 */
	public static boolean stringContainsDelimiter(String string) {
		return Parser.stringContainsDelimiter(string) || string.contains(endDelimiter);
	}
	
	public static boolean stringEndsWithEndDelimiter(String string) {
		return string.endsWith(endDelimiter);
	}
	
	private String commandString = "";
	private HashMap<String, String> parameters;
	
	// --- constructors
	
	/**
	 * Parses a string from server/client
	 * @param parseString which gets parsed to a Command instance
	 */
	public static Command fromString(String parseString) throws IllegalArgumentException {
		// null check
		if (parseString == null) {
			throw new NullPointerException("commandString is null");
		}
		
		// check empty String
		if (parseString.equals("")) {
			throw new IllegalArgumentException("commandstring is empty");
		}
		
		if (!Command.stringEndsWithEndDelimiter(parseString)) {
			throw new IllegalArgumentException("commandstring does not end with end delimiter");
		}
		
		// !!! can throw
		Parser parser = new Parser(parseString.substring(0, parseString.length() - endDelimiter.length()));
		
		ArrayList<Object> arguments = parser.toArrayListOfObjects();
		if (arguments.size() != 2) {
			throw new IllegalArgumentException("too many or too few arguments: " + parseString);
		}
		
		Command command = new Command(" ");
		
		// set commandString
		Object commandStringObject = arguments.get(0);
		command.commandString = Helper.castToString(commandStringObject);
		// !!! can throw
		Helper.throwIllegalErgumentExceptionIfNull("commandString is null", command.commandString);

		// set parameters
		Object parametersObject = arguments.get(1);
		HashMap<String, Object> map = Helper.castToHashMapOfObjects(parametersObject);
		// !!! can throw
		Helper.throwIllegalErgumentExceptionIfNull("parametersObject cast failed", map);
		
		command.parameters = new HashMap<>();
		for (String key : map.keySet()) {
			ArrayList<Object> list = Helper.castToArrayListOfObjects(map.get(key));
			HashMap<String, Object> hashMap = Helper.castToHashMapOfObjects(map.get(key));
			String string = Helper.castToString(map.get(key));
			if (list != null) {
				command.parameters.put(key, new Parser(list).convertToString());
			} else if (hashMap != null) {
				command.parameters.put(key, new Parser(hashMap).convertToString());
			} else if (string != null) {
				command.parameters.put(key, string);
			} else {
				throw new IllegalArgumentException("could not cast object in hash map to ArrayLit, HashMap or String; Object: " + map.get(key));
			}
		}
		
		return command;
	}
	
	public Command(
			String command,
			HashMap<String, String> parameters) throws IllegalArgumentException {
		
		Helper.errorIfNull("One ore more arguments are null", command, parameters);
		
		if (command.equals("")) {
			throw new IllegalArgumentException("commandstring is empty");
		}
		
		// Exception handling
		if (Command.stringContainsDelimiter(command)) {
			throw new IllegalArgumentException("delimiter in command: " + command); }
		for (String key : parameters.keySet()) {
			if (Command.stringContainsDelimiter(key)) {
				throw new IllegalArgumentException("delimiter in one of the keys in parameters HashMap. Key: " + key); }
			if (Command.stringContainsDelimiter(parameters.get(key))) {
				throw new IllegalArgumentException("delimiter in one of the values in parameters HashMap. Value: : " + parameters.get(key)); }
		}
		
		this.commandString = command;
		this.parameters = parameters;
	}
	public Command(String command) throws IllegalArgumentException {
		this(command, new HashMap<String, String>());
	}
	
	// --- setters
	
	public void addParameter(String key, String value) throws IllegalArgumentException {
		// @FIXME check for delimiters in string (currently value can be ObjectConvertible.convertToString())
		// probably a method like Parser.canBeParsed(Object o)
		// or directly use HashMap of Objects instead of Strings
		
		/*if (Command.stringContainsDelimiter(key)) {
			throw new IllegalArgumentException("delimiter in key: " + key); }
		if (Command.stringContainsDelimiter(value)) {
			throw new IllegalArgumentException("delimiter in value: " + value); }*/
		parameters.put(key, value);
	}
	
	// --- getters
	
	public String getCommandString() {
		return commandString;
	}
	
	/**
	 * @param key
	 * @return parameter value for key which can be _NULL_
	 */
	public String getParameterValueForKey(String key) {
		return parameters.get(key);
	}
	
	/**
	 * 
	 * @return the HashMap with all parameters
	 */
	public HashMap<String, String> getParameters(){
		return parameters;
	}
	
	// --- convertToString
	
	public String convertToString() {
		ArrayList<Object> list = new ArrayList<>();
		list.add(commandString);
		list.add(parameters);
		String result = new Parser(list).convertToString() + endDelimiter;
		return result;
	}
	
	/**
	 * @param outputStream through which the command will be send
	 * @throws IOException if "this.convertToString()" could not be written to outputStream
	 */
	public void sendWithOutputStream(OutputStream outputStream) throws IOException {
		// !!! can throw NullPointerException
		Helper.errorIfNull("output Stream is null", outputStream);
		// !!! can throw IOException
		CommandSender.sendCommandWithOutputStream(this, outputStream);
	}
}

File: Robot/Command.swift

import Foundation

/**
 * A class which can send and read Commands over a OutputStream/InputStream
 * 
 * It contains a clientID, commandString and a Dictionary of parameters
 */
public struct Command {
	
	public static let endDelimiter = "%%%%";
	
	/**
	 * @param string which gets checked
	 * @return if string contains one of the delimiters it returns true otherwise false
	 */
	public static func stringContainsDelimiter(_ string: String) -> Bool {
		return Parser.stringContainsDelimiter(string) || string.contains(endDelimiter);
	}
	
	public static func objectContainsDelimiter(_ object: Object) -> Bool {
		switch object {
		case .string(let string):
			return stringContainsDelimiter(string)
		case .array(let array):
			return array.contains(where: objectContainsDelimiter)
		case .dictionary(let dict):
			return dict.keys.contains(where: stringContainsDelimiter) || dict.values.contains(where: objectContainsDelimiter)
		}
	}
	
	public static func stringEndsWithEndDelimiter(_ string: String) -> Bool {
		return string.hasSuffix(endDelimiter);
	}
	
	private var commandString = "";
	private var parameters = [String : Object]();
	
	// --- constructors
	
	/**
	 * Parses a string from server/client
	 * @param parseString which gets parsed to a Command instance
	 */
	public static func fromString(_ parseString: String) throws -> Command {
		
		// check empty String
		if (parseString.equals("")) {
			throw IllegalArgumentException("commandstring is empty");
		}
		
		if (!Command.stringEndsWithEndDelimiter(parseString)) {
			throw IllegalArgumentException("commandstring does not end with end delimiter");
		}
		
		// !!! can throw
		let parser = try Parser(parseString.substring(0, parseString.length() - endDelimiter.length()));
		
		let arguments = parser.toArrayOfObjects()!
		if (arguments.count != 2) {
			throw IllegalArgumentException("too many or too few arguments: " + parseString);
		}

		
		var command = try Command("");
		
		// set commandString
		guard let commandStringObject = Helper.castToString(arguments.get(0)) else {
			throw IllegalArgumentException("commandString is nil");
		}
		command.commandString = commandStringObject

		// set parameters
		guard let map = Helper.castToDictionaryOfObjects(arguments.get(1)) else {
			throw IllegalArgumentException("parametersObject cast failed")
		}
		command.parameters = map

		return command;
	}
	
	public init(
	_ command: String,
	_ parameters: [String : Object]) throws {
		
		if (command == "") {
			throw IllegalArgumentException("command string is empty");
		}
		
		// Exception handling
		if (Command.stringContainsDelimiter(command)) {
			throw  IllegalArgumentException("delimiter in command: " + command); }
		for key in parameters.keys {
			if (Command.stringContainsDelimiter(key)) {
				throw IllegalArgumentException("delimiter in one of the keys in parameters Dictionary. Key: " + key); }
			if (Command.objectContainsDelimiter(parameters.get(key))) {
				throw IllegalArgumentException("delimiter in one of the values in parameters Dictionary. Value: : \(parameters.get(key))");
			}
		}
		
		self.commandString = command;
		self.parameters = parameters;
	}
	public init(
		_ command: String,
		_ parameters: [String : String]) throws {
		
		if (command == "") {
			throw IllegalArgumentException("command string is empty");
		}
		
		// Exception handling
		if (Command.stringContainsDelimiter(command)) {
			throw IllegalArgumentException("delimiter in command: " + command)
		}
		
		self.commandString = command;
		self.parameters = [:]
		for (key, value) in parameters {
			self.addParameter(key, value)
		}
	}
	
	public init(_ command: String) throws {
		try self.init(command, [String : Object]());
	}
	
	// --- setters
	
	public mutating func addParameter(_ key: String, _ value: String) {
		parameters.put(key, .string(value));
	}
	public mutating func addParameter(_ key: String, _ value: Object) {
		parameters.put(key, value);
	}
	
	// --- getters
	
	public func getCommandString() -> String {
		return commandString;
	}
	
	/**
	 * @param key
	 * @return parameter value for key which can be _NULL_
	 */
	public func getParameterValueForKey(_ key: String) -> Object {
		return parameters.get(key)!
	}
	
	/**
	 * 
	 * @return the Dictionary with all parameters
	 */
	func getParameters() -> Dictionary<String, Object> {
		return parameters;
	}
	
	// --- convertToString
	
	public func convertToString() -> String {
		var list = Array<Object>();
		list.add(.string(commandString));
		list.add(.dictionary(parameters));
		let result = Parser(list).convertToString() + Command.endDelimiter;
		return result;
	}
	
	/**
	 * @param outputStream through which the command will be send
	 * @throws IOException if "self.convertToString()" could not be written to outputStream
	 */
	public func sendWithOutputStream(_ outputStream: OutputStream) throws {
		// !!! can throw NullPointerException
		try Helper.errorIfNull("output Stream is nil", outputStream);
		// !!! can throw IOException
		try CommandSender.sendCommandWithOutputStream(self, outputStream);
	}
}
